### PR TITLE
Scheduled Digests

### LINKED ISSUE #418: [P4] Scheduled Digests
Scheduled alert summaries(daily/weekly)

### PR DESCRIPTION
**Pull request by kunalsz**
Resolves #418 

## Changes Made:
- Adds a new celery task to send scheduled alert summaries (daily and weekly)
- Adds a new method `send_scheduled_summary` in alerters
- New template for alert summaries `alert_template_summary.jinja`

### CODE REVIEW

File: buffalogs/impossible_travel/tasks.py [hunk 1]
```diff
@@ -113,3 +113,55 @@ def notify_alerts():
...
+    )
+
+    # Get alerts within the defined time range
+    start_date = task_settings.start_date
+    end_date = task_settings.end_date
+    alerts = Alert.objects.filter(created__range=(start_date, end_date))
+
+    total_alerts = alerts.count()
+
+    # Group alerts by user {user: {alert:count}}
```
Comment (Lorygold) [inline buffalogs/impossible_travel/tasks.py:146]: I wasn't sure that this block of code ran correctly, so I tested it (see image below).

image

The issue is that here you initialize `user_breakdown[alert.user.username]` as a **list**, but then you treat it like a **dictionary** when doing `user_breakdown[alert.user.username][alert.name] += 1`. That’s why it doesn’t work as expected.

A neat way to solve this is by using a `defaultdict` with a `lambda` that creates another `defaultdict(int)`. For example:
```
from collections import defaultdict

user_breakdown = defaultdict(lambda: defaultdict(int))
for alert in alerts:
    user_breakdown[alert.user.username][alert.name] += 1
```
This way each new user automatically gets a dictionary of alert counts, and each alert type starts at 0 — so the increment will always work.

File: buffalogs/impossible_travel/tasks.py [hunk 2]
```diff
@@ -113,3 +113,55 @@ def notify_alerts():
...
+    user_breakdown = {}
+    for alert in alerts:
```
Comment (Lorygold) [inline buffalogs/impossible_travel/tasks.py:148]: same here

One more tip: issues like this are often caught when you have good test coverage. The more cases you test (different users, different alert types, edge cases with no alerts, etc.), the quicker these bugs show up. It’s usually **better to have one extra test than one too few** :relaxed:

File: buffalogs/impossible_travel/alerting/slack_alerting.py [hunk 3]
```diff
@@ -49,6 +49,28 @@ def send_message(self, alert):
...
+        summary_msg = {
+            "attachments": [
+                {
+                    "title": summary_title,
+                    "text": summary_description,
+                    "color": "#ff0000",
+                }
+            ]
+        }
+        resp = requests.post(self.webhook_url, json=summary_msg, headers={"Content-Type": "application/json"})
```
Comment (Lorygold) [inline buffalogs/impossible_travel/alerting/slack_alerting.py:71]: This is better to insert into a try-except block, logging the error:
```
import logging

logger = logging.getLogger(__name__)

try:
    resp = requests.post(self.webhook_url, json=summary_msg)
    resp.raise_for_status()
except requests.RequestException as e:
    logger.error(f"Slack summary notification failed: {e}")
```

File: buffalogs/impossible_travel/tasks.py (hunk 2 above)
Comment (kunalsz) [reply buffalogs/impossible_travel/tasks.py:148]: You are right. I'll keep that in mind !

**Comment by kunalsz** [comment]
@Lorygold I have made the necessary changes and refactors, is it good to go ?
